src/browser/wisp_network.js

```diff
@@ -0,0 +1,302 @@
+{
+    if(this.wispws) {
+        this.wispws.onmessage = null;
+        this.wispws.onclose = null;
+        this.wispws.close();
+        this.wispws = null;
+    }
+};
+
+/**
+ * @param {Uint8Array} data
+ */
+WispNetworkAdapter.prototype.send = function(data)
+{
+    let packet = {};
+    parse_eth(data, packet);
+
+    if(packet.tcp) {
+        let reply = {};
+        reply.eth = { ethertype: ETHERTYPE_IPV4, src: this.router_mac, dest: packet.eth.src };
+        reply.ipv4 = {
+            proto: IPV4_PROTO_TCP,
+            src: packet.ipv4.dest,
+            dest: packet.ipv4.src
+        };
+
+        let tuple = [
+            packet.ipv4.src.join("."),
+            packet.tcp.sport,
+            packet.ipv4.dest.join("."),
+            packet.tcp.dport
+        ].join(":");
+
+        if(packet.tcp.syn) {
+            if(this.tcp_conn[tuple]) {
+                dbg_log("SYN to already opened port", LOG_FETCH);
+            }
+            const tcp_conn = new TCPConnection();
+
+            tcp_conn.state = TCP_STATE_SYN_RECEIVED;
+            tcp_conn.net = this;
+            tcp_conn.tuple = tuple;
+            tcp_conn.stream_id = this.last_stream++;
+            this.tcp_conn[tuple] = tcp_conn;
+
+            tcp_conn.on_data = (data) => {
+                if(data.length !== 0) {
+                    this.send_wisp_frame({
+                        type: "DATA",
+                        stream_id: tcp_conn.stream_id,
+                        data: data
+                    });
+                }
+            };
+
+            this.send_wisp_frame({
+                type: "CONNECT",
+                stream_id: tcp_conn.stream_id,
+                hostname: packet.ipv4.dest.join("."),
+                port: packet.tcp.dport,
+                data_callback: (data) => {
+                    tcp_conn.write(data);
+                },
+                close_callback: (data) => {
+                    tcp_conn.close();
+                }
+            });
+
+            tcp_conn.accept(packet);
+            return;
+        }
+
+        if(!this.tcp_conn[tuple]) {
+            dbg_log(`I dont know about ${tuple}, so restting`, LOG_FETCH);
+            let bop = packet.tcp.ackn;
+            if(packet.tcp.fin || packet.tcp.syn) bop += 1;
+            reply.tcp = {
+                sport: packet.tcp.dport,
+                dport: packet.tcp.sport,
+                seq: bop,
+                ackn: packet.tcp.seq + (packet.tcp.syn ? 1: 0),
+                winsize: packet.tcp.winsize,
+                rst: true,
+                ack: packet.tcp.syn
+            };
+            this.receive(make_packet(reply));
+            return;
+        }
+
+        this.tcp_conn[tuple].process(packet);
+    }
+
+    if(packet.arp && packet.arp.oper === 1 && packet.arp.ptype === ETHERTYPE_IPV4) {
+        arp_whohas(packet, this);
+    }
+
+    if(packet.dns) {
+        // TODO: remove when this wisp client supports udp
```

Comment:
FYI, I would like to avoid DOH (behind a switch that is disabled by default), in order to avoid leaking hostnames to cloudflare. However, since this wisp implementation doesn't have UDP yet, I'll leave it with this comment for now.

Reply:
Understandable, I'm a bit busy this week but I'll see if I can make out time in the near future to add in UDP